zio-json/shared/src/main/scala/zio/json/uuid/UUIDParser.scala

```diff
@@ -0,0 +1,154 @@
+
+// A port of https://github.com/openjdk/jdk/commit/ebadfaeb2e1cc7b5ce5f101cd8a539bc5478cf5b
+private[json] object UUIDParser {
+  private val HexToDecimal: Array[Byte] = {
+    val ns = Array.ofDim[Byte](256)
+    ns('0') = 0
+    ns('1') = 1
+    ns('2') = 2
+    ns('3') = 3
+    ns('4') = 4
+    ns('5') = 5
+    ns('6') = 6
+    ns('7') = 7
+    ns('8') = 8
+    ns('9') = 9
+    ns('A') = 10
+    ns('B') = 11
+    ns('C') = 12
+    ns('D') = 13
+    ns('E') = 14
+    ns('F') = 15
+    ns('a') = 10
+    ns('b') = 11
+    ns('c') = 12
+    ns('d') = 13
+    ns('e') = 14
+    ns('f') = 15
+    ns
+  }
+
+  def unsafeParse(input: String): java.util.UUID =
+    if (input.length == 36) {
+      val ch1 = input.charAt(8)
+      val ch2 = input.charAt(13)
+      val ch3 = input.charAt(18)
+      val ch4 = input.charAt(23)
+
+      if (ch1 == '-' && ch2 == '-' && ch3 == '-' && ch4 == '-') {
+        val msb1 = parseNibbles(input, 0)
+        val msb2 = parseNibbles(input, 4)
+        val msb3 = parseNibbles(input, 9)
+        val msb4 = parseNibbles(input, 14)
+        val lsb1 = parseNibbles(input, 19)
+        val lsb2 = parseNibbles(input, 24)
+        val lsb3 = parseNibbles(input, 28)
+        val lsb4 = parseNibbles(input, 32)
+        if ((msb1 | msb2 | msb3 | msb4 | lsb1 | lsb2 | lsb3 | lsb4) >= 0) {
+          new java.util.UUID(
+            msb1 << 48 | msb2 << 32 | msb3 << 16 | msb4,
+            lsb1 << 48 | lsb2 << 32 | lsb3 << 16 | lsb4
+          )
+        } else unsafeParseExtended(input)
+      } else unsafeParseExtended(input)
+    } else unsafeParseExtended(input)
+
+  // A nibble is 4 bits
+  @nowarn("msg=implicit numeric widening")
+  private def parseNibbles(input: String, position: Int): Long = {
+    val ns  = HexToDecimal
+    val ch1 = input.charAt(position)
+    val ch2 = input.charAt(position + 1)
+    val ch3 = input.charAt(position + 2)
+    val ch4 = input.charAt(position + 3)
+
+    if ((ch1 | ch2 | ch3 | ch4) > 0xff) -1L
+    else ns(ch1) << 12 | ns(ch2) << 8 | ns(ch3) << 4 | ns(ch4)
+  }
+
+  private def unsafeParseExtended(name: String): java.util.UUID = {
+    val len = name.length
+    if (len > 36) throw new IllegalArgumentException("UUID string too large")
+    val dash1 = name.indexOf('-', 0)
+    val dash2 = name.indexOf('-', dash1 + 1)
+    val dash3 = name.indexOf('-', dash2 + 1)
+    val dash4 = name.indexOf('-', dash3 + 1)
+    val dash5 = name.indexOf('-', dash4 + 1)
+
+    // For any valid input, dash1 through dash4 will be positive and dash5 will be negative,
+    // but it's enough to check dash4 and dash5:
+    // - if dash1 is -1, dash4 will be -1
+    // - if dash1 is positive but dash2 is -1, dash4 will be -1
+    // - if dash1 and dash2 is positive, dash3 will be -1, dash4 will be positive, but so will dash5
+    if (dash4 < 0 || dash5 >= 0) throw new IllegalArgumentException("Invalid UUID string: " + name)
+
+    var mostSigBits = parseLong(name, 0, dash1, 16) & 0xffffffffL
```

Comment:
Possible extra bits are masked here. Shouldn't they trigger a parsing error instead?
```
$ sbt -java-home /usr/lib/jvm/openjdk-17 console
[info] welcome to sbt 1.5.3 (Oracle Corporation Java 17-ea)
...
[info] Starting scala interpreter...
Welcome to Scala 2.13.6 (OpenJDK 64-Bit Server VM, Java 17-ea).
Type in expressions for evaluation. Or try :help.

scala> java.util.UUID.fromString("FFFFFFFFFFFFFFF-0-0-0-0")
val res0: java.util.UUID = ffffffff-0000-0000-0000-000000000000
```

Reply:
How would I go about detecting and fixing this? I'm not sure how to approach this, I would appreciate your insight!

Reply:
Before accumulation we can mask and check if extra bits are non-zero.

Reply:
Something like this perhaps?
```scala
val mask1: Long = 0xFFFFFFFFFFFFL  // selected 12 hex digits to cover longest section in UUID
val section1 = 32    // 8 * (4 bits per hex)
val section234 = 16  // 4 * (4 bits per hex)
val section5 = 48    // 12 * (4 bits per hex)

def checkExtraBits(input: Long, section: Int): Long = {
  if (((input >> section) & mask1) > 0) throw new NumberFormatException(s"Extra bits detected in section: ${input.toHexString}")
  else input
}
```

With the example provided:
```scala
checkExtraBits(0xFFFFFFFFFFFFFFFL, section1).toHexString
//  0xFFFFFFFFFFFFFFFL >> 32 (section1)
//  0xFFFFFFF
//  0xFFFFFFF & 0xFFFFFFFFFFFFL (mask1)
//  0xFFFFFFF > 1
// java.lang.NumberFormatException: Extra bits detected in section: fffffffffffffff
```

Thank you for your help and insight 🙏🏽

Reply:
Oops, I'm dumb, I just realized you can do this (which I think is what you meant earlier):
```scala
val maskFirst8 = 0xFFFFFFFF0000000L
val invalid =    0xFFFFFFFFFFFFFFFL
val valid =      0x123E4567L

maskFirst8 & invalid // > 0
maskFirst8 & valid   // == 0
```

I believe we would need to have masks for each section right? (8-4-4-4-12)

Reply:
I believe this should be done now so we zero mask and check if we have any leftover bits for each section. Please let me know if the implementation is alright with you.